File: 3 КУРС/2 семестр/Java/VoiceAssistent/MainActivity.java
package com.example.voiceassistent;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.util.Consumer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.Parcelable;
import android.speech.tts.TextToSpeech;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.content.Context;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private static final String URL = "http://mirkosmosa.ru/holiday/2021";
    private static final String APP_PREFERENCES = "mysettings";
    private boolean isLight = true;
    private String THEME = "THEME";
    protected Button sendButton;
    protected EditText questionText;
    protected TextView chatWindow;
    SharedPreferences sPref;
    TextToSpeech textToSpeech;
    protected RecyclerView chatMessageList;
    protected MessageListAdapter messageListAdapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_main);
        sPref = getSharedPreferences(APP_PREFERENCES,MODE_PRIVATE);
        messageListAdapter = new MessageListAdapter();
        chatMessageList=findViewById(R.id.chatMessageList);
        chatMessageList.setLayoutManager(new LinearLayoutManager(this));
        chatMessageList.setAdapter(messageListAdapter);
        super.onCreate(savedInstanceState);
        sendButton = findViewById(R.id.sendButton);
        questionText = findViewById(R.id.questionField);
        isLight = sPref.getBoolean(THEME, true);
        if (isLight) getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
        else getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES);
        textToSpeech = new TextToSpeech(this, new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {
                if (status == TextToSpeech.SUCCESS) {
                    textToSpeech.setLanguage(new Locale("ru"));
                }
                else chatWindow.append("Can't text to speech");
            }
        });
        sendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onSend();
            }
        });
    }
    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        String[] messageList= new String[messageListAdapter.messageList.size()];
        String [] dateList =  new String[messageListAdapter.messageList.size()];
        boolean[] sendList = new boolean[messageListAdapter.messageList.size()];
        for (int i=0; i< messageList.length;++i){
            messageList[i] = messageListAdapter.messageList.get(i).text;
            sendList[i] = messageListAdapter.messageList.get(i).isSend;
            dateList[i] = (String.valueOf(messageListAdapter.messageList.get(i).date.getTime()));
        }
        outState.putStringArray("messageList",messageList);
        outState.putStringArray("dateList",dateList);
        outState.putBooleanArray("sendList", sendList);

    }
    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        String[] tempMessage = savedInstanceState.getStringArray("messageList");
        String[] tempDate = savedInstanceState.getStringArray("dateList");
        boolean[] tempSend = savedInstanceState.getBooleanArray("sendList");
        for (int i=0; i<tempMessage.length; ++i) {
            messageListAdapter.messageList.add(new Message(tempMessage[i], tempSend[i], tempDate[i]));
        }
    }
    protected void onSend(){
        String text = questionText.getText().toString();
        AI_MAP temp = new AI_MAP();
        temp.get(text, s -> {
            messageListAdapter.messageList.add(new Message(text, true));
            messageListAdapter.messageList.add(new Message(s, false));
            messageListAdapter.notifyDataSetChanged();
            chatMessageList.scrollToPosition(messageListAdapter.messageList.size()-1);
            textToSpeech.speak(s, TextToSpeech.QUEUE_FLUSH,null, null );
            questionText.setText("");
        });
    }

    @Override
    protected void onStop() {
        SharedPreferences.Editor editor = sPref.edit();
        editor.putBoolean(THEME, isLight);
        editor.apply();
        super.onStop();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.day_settings:
                isLight=true;
                getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                break;
            case R.id.night_settings:
                isLight=false;
                getDelegate().setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public SharedPreferences getSharedPreferences(String name, int mode) {
        return super.getSharedPreferences(name, mode);
    }
}

File: 3 КУРС/2 семестр/Учебная практика/kotlin/kotlin/app/src/main/java/com/example/kotlinapp/MainActivity.kt
package com.example.kotlinapp

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.content.SharedPreferences
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.os.Build
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.TextToSpeech.OnInitListener
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.kotlinapp.Database.DBHelper
import com.example.kotlinapp.Database.DBHelper.Companion.FIELD_DATE
import com.example.kotlinapp.Database.DBHelper.Companion.FIELD_MESSAGE
import com.example.kotlinapp.Database.DBHelper.Companion.FIELD_SEND
import com.example.kotlinapp.Database.DBHelper.Companion.TABLE_MESSAGES
import com.example.kotlinapp.Database.MessageEntity
import com.example.kotlinapp.MessageService.Message
import com.example.kotlinapp.MessageService.MessageListAdapter
import java.io.IOException
import java.text.ParseException
import java.util.*
import java.util.function.Consumer


open class MainActivity : AppCompatActivity() {
    protected var sendButton: Button? = null
    protected var questionText: EditText? = null
    protected var textToSpeech: TextToSpeech? = null
    protected var chatMessageList: RecyclerView? = null
    protected var messageListAdapter: MessageListAdapter? = null
    protected var sPref: SharedPreferences? = null
    private var isLight = true
    private val THEME = "THEME"
    var dBHelper: DBHelper? = null
    lateinit var database: SQLiteDatabase
    var cursor: Cursor? = null

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    override fun onCreate(savedInstanceState: Bundle?) {
        sPref = getSharedPreferences(
            APP_PREFERENCES,
            Context.MODE_PRIVATE
        )
        isLight = sPref?.getBoolean(THEME, true)!!
        if (isLight == false) delegate.localNightMode = AppCompatDelegate.MODE_NIGHT_YES
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        sendButton = findViewById(R.id.sendButton)
        questionText = findViewById(R.id.questionField)
        chatMessageList = findViewById(R.id.chatMessageList)
        messageListAdapter = MessageListAdapter()
        chatMessageList?.layoutManager = LinearLayoutManager(this)
        chatMessageList?.adapter = messageListAdapter
        dBHelper = DBHelper(this)
        database = dBHelper!!.writableDatabase


        var str: ArrayList<Message>
        if (savedInstanceState != null) {
            str = (savedInstanceState.getSerializable("key") as ArrayList<Message>?)!!
            for (i in str!!.indices) {
                if (i % 2 == 0)
                    messageListAdapter!!.messageList.add(str[i])
                if (i % 2 != 0)
                    messageListAdapter!!.messageList.add(str[i])
            }
            messageListAdapter!!.notifyDataSetChanged()
        }

        if (savedInstanceState == null) {
            val cursor = database.query(TABLE_MESSAGES, null, null, null, null, null, null)
            if (cursor.moveToFirst()) {
                val messageIndex = cursor.getColumnIndex(FIELD_MESSAGE)
                val dateIndex = cursor.getColumnIndex(FIELD_DATE)
                val sendIndex = cursor.getColumnIndex(FIELD_SEND)
                do {
                    val entity = MessageEntity(
                        cursor.getString(messageIndex),
                        cursor.getString(dateIndex), cursor.getInt(sendIndex)
                    )
                    val message = Message(entity)
                    if (message.date == null || message.text == null || message.isSend == null) {
                        //
                    } else {
                        messageListAdapter!!.messageList.add(message)
                    }
                } while (cursor.moveToNext())
            }
            cursor.close()
        }

        textToSpeech = TextToSpeech(applicationContext,
            OnInitListener { i ->
                if (i != TextToSpeech.ERROR) {
                    textToSpeech!!.language = Locale("ru")
                }
            })


        sendButton?.setOnClickListener(View.OnClickListener {
            try {
                onSend()
            } catch (e: ParseException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val s =
            messageListAdapter!!.messageList as ArrayList<Message>
        outState.putSerializable("key", s)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onDestroy() {
        if (textToSpeech != null) {
            textToSpeech!!.stop()
            textToSpeech!!.shutdown()
        }
        database.close()
        dBHelper!!.close()
        super.onDestroy()
    }

    override fun onStop() {
        val editor = sPref!!.edit()
        editor.putBoolean(THEME, isLight)
        editor.apply()
        database!!.delete(TABLE_MESSAGES, null, null)
        for (i in 0 until messageListAdapter!!.messageList.size) {
            val entity = MessageEntity(messageListAdapter!!.messageList[i])
            val contentValues = ContentValues()
            contentValues.put(DBHelper.FIELD_MESSAGE, entity.text)
            contentValues.put(DBHelper.FIELD_SEND, entity.isSend)
            contentValues.put(DBHelper.FIELD_DATE, entity.date)
            database!!.insert(TABLE_MESSAGES, null, contentValues)
        }
        super.onStop()
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.day_settings -> {
                delegate.localNightMode = AppCompatDelegate.MODE_NIGHT_NO
                isLight = true
            }
            R.id.night_settings -> {
                delegate.localNightMode = AppCompatDelegate.MODE_NIGHT_YES
                isLight = false
            }
            R.id.clear_dialog -> {
                messageListAdapter!!.messageList.clear();
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    @SuppressLint("NewApi")
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Throws(
        ParseException::class,
        IOException::class
    )
    protected open fun onSend() {
        val text = questionText!!.text.toString()
        AI.getAnswer(text, Consumer<String> { s ->
            Log.i("WEATHER", "MA: $s")
            textToSpeech!!.speak(s, TextToSpeech.QUEUE_FLUSH, null, null)
            messageListAdapter!!.messageList.add(Message(text, true))
            messageListAdapter!!.messageList.add(Message(s, false))
            messageListAdapter!!.notifyDataSetChanged()
            chatMessageList!!.scrollToPosition(messageListAdapter!!.messageList.size - 1)
            questionText!!.setText("")
        })
    }

    companion object {
        const val APP_PREFERENCES = "mysettings"
    }
}